### PR TITLE
[extractor/japandiet] Add extractors

### PR DESCRIPTION
**Pull request by Lesmiscore**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

This adds support for proceedings live/record for House of Representatives (Shugiin) and House of Councillors (Sangiin) of Japan.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/japandiet.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+class ShugiinItvBaseIE(InfoExtractor):
+    @classmethod
+    def _find_rooms(cls, webpage):
+        return [{
+            '_type': 'url',
+            'id': x.group(1),
+            'title': clean_html(x.group(2)).strip(),
+            'url': smuggle_url(f'https://www.shugiintv.go.jp/jp/index.php?room_id={x.group(1)}', x.groups()),
+            'ie_key': ShugiinItvLiveIE.ie_key(),
+        } for x in re.finditer(r'<a\s+href=".+?\?room_id=(room\d+)"\s*class="play_live".+?class="s12_14">(.+?)</td>', webpage)]
```
Comment (pukkandan) [inline yt_dlp/extractor/japandiet.py:25]:
```suggestion
        } for x in re.finditer(r'<a\s+href="[^"]+\?room_id=(room\d+)"\s*class="play_live".+?class="s12_14">(.+?)</td>', webpage)]
```

The regex looks fragile. Though I haven't looked at the page, so not sure if there's a better way

File: yt_dlp/extractor/japandiet.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        duration = 0
+        if secs:
+            duration += float(secs)
+        if mins:
+            duration += float(mins) * 60
+        if hours:
+            duration += float(hours) * 60 * 60
+        if days:
+            duration += float(days) * 24 * 60 * 60
+        return duration
```
Comment (pukkandan) [inline yt_dlp/extractor/japandiet.py:68]: Are these standard enough to go into parse_duration/unified_strdate?

File: yt_dlp/extractor/japandiet.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+class ShugiinItvLiveIE(ShugiinItvBaseIE):
+    _VALID_URL = r'https?://(?:www\.)?shugiintv\.go\.jp/(?:jp|en)(?:/index\.php)?$'
+    IE_DESC = '衆議院インターネット審議中継'
+
+    @classmethod
+    def suitable(cls, url):
+        return super().suitable(url) and not any(x.suitable(url) for x in (ShugiinItvLiveRoomIE, ShugiinItvVodIE))
+
+    def _real_extract(self, url):
+        self.report_warning('Listing up all running proceedings as of now. To specify one proceedings to record, use link direct from the website.')
```
Comment (pukkandan) [inline yt_dlp/extractor/japandiet.py:80]:
```suggestion
        self.to_screen(
            'Downloading all running proceedings. To specify one proceeding, use direct link from the website')
```

File: yt_dlp/extractor/japandiet.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+        if mins:
+            duration += float(mins) * 60
+        if hours:
+            duration += float(hours) * 60 * 60
+        if days:
+            duration += float(days) * 24 * 60 * 60
+        return duration
+
+
+class ShugiinItvLiveIE(ShugiinItvBaseIE):
```
Comment (pukkandan) [inline yt_dlp/extractor/japandiet.py:71]: no test?

File: yt_dlp/extractor/japandiet.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+    def _real_extract(self, url):
+        url, smug = unsmuggle_url(url)
+        if smug:
+            room_id, title = smug
+        else:
+            room_id = self._match_id(url)
+            webpage = self._download_webpage(
+                'https://www.shugiintv.go.jp/jp/index.php', room_id,
+                encoding='euc-jp', note='Looking up for the title')
```
Comment (pukkandan) [inline yt_dlp/extractor/japandiet.py:99]:
```suggestion
                encoding='euc-jp', note='Looking up stream title')
```

File: yt_dlp/extractor/japandiet.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(
+            f'https://www.shugiintv.go.jp/jp/index.php?ex=VL&media_type=&deli_id={video_id}', video_id,
+            encoding='euc-jp')
+
+        m3u8_url = self._search_regex(
+            r'id="vtag_src_base_vod"\s*value="(http.+?\.m3u8)"', webpage, 'm3u8 url')
+        m3u8_url = re.sub(r'^http://', 'https://', m3u8_url)
```
Comment (pukkandan) [inline yt_dlp/extractor/japandiet.py:140]: is this necessary?

File: yt_dlp/extractor/japandiet.py [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+        # NOTE: chapters are sparse, because of how the website serves the video
+        chapters = []
+        for chp in re.finditer(r'<A\s+HREF=".+?php\?.+?&deli_id=\d+&time=([\d\.]+)"\s*class="play_vod">(?!<img)(.+)</[Aa]>', webpage):
+            chapters.append({
+                'title': clean_html(chp.group(2)).strip(),
+                'start_time': float_or_none(chp.group(1).strip()),
+            })
+        # the exact duration for the last chapter is unknown! (we can get at most minutes of granularity)
+        for idx in range(len(chapters) - 1):
+            chapters[idx]['end_time'] = chapters[idx + 1]['start_time']
```
Comment (pukkandan) [inline yt_dlp/extractor/japandiet.py:162]: core code handles this

File: yt_dlp/extractor/japandiet.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+            (r'<td\s+align="left">(.+)\s*\(\d+分\)',
+             r'<TD.+?<IMG\s*src=".+?/spacer\.gif".+?height="15">(.+?)<IMG'), webpage, 'title', fatal=False)
+
+        release_date = self._parse_japanese_date(self._html_search_regex(
+            r'開会日</td>\s*<td.+?/td>\s*<TD>(.+?)</TD>',
+            webpage, 'title', fatal=False))
+
+        # NOTE: chapters are sparse, because of how the website serves the video
+        chapters = []
+        for chp in re.finditer(r'<A\s+HREF=".+?php\?.+?&deli_id=\d+&time=([\d\.]+)"\s*class="play_vod">(?!<img)(.+)</[Aa]>', webpage):
```
Comment (pukkandan) [inline yt_dlp/extractor/japandiet.py:155]:
```suggestion
        for chp in re.finditer(r'(?i)<a\s+href="[^"]+php\?[^"]+?&deli_id=\d+&time=([\d.]+)"\s*class="play_vod">(?!<img)(.+?)</a>', webpage):
```
Capture the whole URL and parse query later imo. Currently this breaks if order of query params change

File: yt_dlp/extractor/japandiet.py [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+            })
+        # the exact duration for the last chapter is unknown! (we can get at most minutes of granularity)
+        for idx in range(len(chapters) - 1):
+            chapters[idx]['end_time'] = chapters[idx + 1]['start_time']
+
+        last_tr = re.findall(r'(?s)<TR\s*class="s14_24">(.+?)</TR>', webpage)[-1]
+        if last_tr and chapters:
+            last_td = re.findall(r'<TD.+?</TD>', last_tr.group(0))[-1]
+            if last_td:
+                chapters[-1]['end_time'] = chapters[-1]['start_time'] + self._parse_japanese_duration(clean_html(last_td.group(0)))
```
Comment (pukkandan) [inline yt_dlp/extractor/japandiet.py:168]: set this to `duration` instead

File: yt_dlp/extractor/japandiet.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+
+    def _real_extract(self, url):
+        self.report_warning('Listing up all running proceedings as of now. To specify one proceedings to record, use link direct from the website.')
+        webpage = self._download_webpage(
+            'https://www.shugiintv.go.jp/jp/index.php', None,
+            encoding='euc-jp')
+        return self.playlist_result(self._find_rooms(webpage))
+
+
+class ShugiinItvLiveRoomIE(ShugiinItvBaseIE):
```
Comment (pukkandan) [inline yt_dlp/extractor/japandiet.py:87]: add a test even if it'll become outdated

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/japandiet.py (hunk 4 above)
Comment (Lesmiscore) [reply yt_dlp/extractor/japandiet.py:71]: whoops, let me do this tomorrow since I'm busy today

File: yt_dlp/extractor/japandiet.py (hunk 10 above)
Comment (Lesmiscore) [reply yt_dlp/extractor/japandiet.py:88]: whoops (again)

File: yt_dlp/extractor/japandiet.py (hunk 6 above)
Comment (Lesmiscore) [reply yt_dlp/extractor/japandiet.py:141]: We'll get the error page without it

File: yt_dlp/extractor/japandiet.py (hunk 9 above)
Comment (Lesmiscore) [reply yt_dlp/extractor/japandiet.py:169]: The `duration` key in the last chapter?

File: yt_dlp/extractor/japandiet.py (hunk 2 above)
Comment (Lesmiscore) [reply yt_dlp/extractor/japandiet.py:68]: Do you wish them to be messy by adding thousands of formatting strings from the world? (also imo it's hard to merge into the existing functions in utils)

File: yt_dlp/extractor/japandiet.py (hunk 9 above)
Comment (pukkandan) [reply yt_dlp/extractor/japandiet.py]: No, duration of the video. Core code will copy it into last chapter + we'll get duration metadata too

Comment (Lesmiscore) [reply yt_dlp/extractor/japandiet.py]: image
The "duration" being extracted is this field, which is titled "duration of this segment" ("所要時間" in the image) above it
Also, the duration mentioned in the webpage is incorrect - it doesn't include the blank after and before chapters

Comment (pukkandan) [reply yt_dlp/extractor/japandiet.py]: gotcha. Pls add a comment explaining this

File: yt_dlp/extractor/japandiet.py (hunk 2 above)
Comment (pukkandan) [reply yt_dlp/extractor/japandiet.py:69]: hm.. I get ur point. Ig we can move it only if other sites need it in future

btw, I'd suggest moving this to module level

File: yt_dlp/extractor/japandiet.py [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+        duration = 0
+        if secs:
+            duration += float(secs)
+        if mins:
+            duration += float(mins) * 60
+        if hours:
+            duration += float(hours) * 60 * 60
+        if days:
+            duration += float(days) * 24 * 60 * 60
+        return duration
```
Comment (pukkandan) [inline yt_dlp/extractor/japandiet.py:69]:
```suggestion
        mobj = re.search(r'(?:(\d+)日間?)?(?:(\d+)時間?)?(?:(\d+)分)?(?:(\d+)秒)?', re.sub(r'[\s\u3000]+', '', text or ''))
        if not mobj:
            return
        days, hours, mins, secs = [int_or_none(x, default=0) for x in mobj.groups()]
        return secs + mins * 60 + hours * 60 * 60 + days * 24 * 60 * 60
```

File: yt_dlp/extractor/japandiet.py [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+
+class ShugiinItvLiveIE(ShugiinItvBaseIE):
+    _VALID_URL = r'https?://(?:www\.)?shugiintv\.go\.jp/(?:jp|en)(?:/index\.php)?$'
+    IE_DESC = '衆議院インターネット審議中継'
+
+    _TESTS = [{
+        'url': 'https://www.shugiintv.go.jp/jp/index.php',
+        # expect at least one proceedings is running
+        'playlist_mincount': 1,
+        'only_matching': True,
```
Comment (pukkandan) [inline yt_dlp/extractor/japandiet.py:80]: why only_matching? The playlist metadata should be static

File: yt_dlp/extractor/japandiet.py [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+    }]
+
+    @classmethod
+    def suitable(cls, url):
+        return super().suitable(url) and not any(x.suitable(url) for x in (ShugiinItvLiveRoomIE, ShugiinItvVodIE))
```
Comment (pukkandan) [inline yt_dlp/extractor/japandiet.py:85]: We really need to generalize this pattern 🤔

File: yt_dlp/extractor/japandiet.py [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+
+    def _real_extract(self, url):
+        self.to_screen(
+            'Downloading all running proceedings. To specify one proceeding, use direct link from the website')
+        webpage = self._download_webpage(
+            'https://www.shugiintv.go.jp/jp/index.php', None,
+            encoding='euc-jp')
+        return self.playlist_result(self._find_rooms(webpage))
```
Comment (pukkandan) [inline yt_dlp/extractor/japandiet.py:93]: I would recommend adding either an id or title. Maybe `all proceedings`?

File: yt_dlp/extractor/japandiet.py [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+        'info_dict': {
+            'id': 'room11',
+            'title': '外務委員会',
+        },
+        'skip': 'this runs for a time and not every day',
+    }]
+
+    def _real_extract(self, url):
+        url, smug = unsmuggle_url(url)
+        if smug and smug.get('g'):
```
Comment (pukkandan) [inline yt_dlp/extractor/japandiet.py:118]:
```suggestion
        url, smug = unsmuggle_url(url, default={})
        if smug.get('g'):
```

File: yt_dlp/extractor/japandiet.py [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+            room_id, title = smug['g']
+        else:
+            room_id = self._match_id(url)
+            webpage = self._download_webpage(
+                'https://www.shugiintv.go.jp/jp/index.php', room_id,
+                encoding='euc-jp', note='Looking up stream title')
```
Comment (pukkandan) [inline yt_dlp/extractor/japandiet.py:124]: maybe this should be cached in RAM? To speed up `-a`

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/japandiet.py (hunk 12 above)
Comment (Lesmiscore) [reply yt_dlp/extractor/japandiet.py]: added

**Pull request merged**